api/admin/controller/self_tests.py

```diff
@@ -20,14 +20,22 @@ def process_get(self, identifier):
         if isinstance(item, ProblemDetail):
             return item
         info = self.get_info(item)
-        info["self_test_results"] = self._get_prior_test_results(item)
+        protocol_class = None
+        if hasattr(self, "_find_protocol_class"):
+            protocol_class = self._find_protocol_class(item)
+        info["self_test_results"] = self._get_prior_test_results(item, protocol_class)
         return dict(self_test_results=info)
 
     def process_post(self, identifier):
         item = self.look_up_by_id(identifier)
         if isinstance (item, ProblemDetail):
             return item
         value = self.run_tests(item)
-        if (value):
+        if value and isinstance(value, ProblemDetail):
+            return value
+        elif value:
             return Response(_("Successfully ran new self tests"), 200)
-        return FAILED_TO_RUN_SELF_TESTS
+        return FAILED_TO_RUN_SELF_TESTS.detailed(_(
+            "Failed to run self tests for this %s."
+            %(self.type)
+        ))
```

Comment:
A couple minor things about `_()`:

* The string interpolation needs to go outside of the `_()` call, because the string in the `_()` call is what gets translated.

* It's better if the whole thing is on one line, because we use an automated tool to grab out the strings to be translated.

So:

```
return FAILED_TO_RUN_SELF_TESTS.detailed(
    _("Failed to run self tests for this %s.") % self.type
)
```

Even this isn't perfect because `self.type` also needs to be translated, but this will at least make us notice that there's some translation happening here.

Reply:
I would do 
```
return FAILED_TO_RUN_SELF_TESTS.detailed(
    _("Failed to run self tests for this %(type)s.", self.type)
)
```

and then also use `_()` on the type names where those are defined.